src/GenSequence.ts

```diff
@@ -282,6 +286,33 @@ export function first<T>(fn: (t: T) => boolean, defaultValue: T, i: Iterable<T>)
     return defaultValue;
 }
 
+export function max<T>(i: Iterable<T>): Maybe<T> {
+    let sawValue: boolean = false;
+    const value: Maybe<T> = reduce((p: T, c: T) => {
+        if (!sawValue) {
+            sawValue = true;
+        }
+        
+        if (isNonValue(p) || c > p) {
+            return c;
+        }
+
+        return p;
+    },
+    undefined, 
+    i);
+
+    if (!sawValue) {
+        throw new Error("the sequence is empty.");
+    }
+    
+    if (isNonValue(value)) {
+        return undefined;
+    }
+
+    return value;
+}
+
```

Comment:
My vote in this case is to keep it simple. `.max()` should do only one thing.

```typescript
export function max<T>(i: Iterable<T>): Maybe<T> {
    return reduce((p: T, c: T) => c > p ? c : p, undefined, i);
}
```

Thoughts:
* Empty lists: the *max* on an empty list is really just `undefined` and not an error. I can see why it might be tempting to throw an exception after looking at `Array.reduce()`.  In the case of `Array.reduce()`, the exception is only thrown in the cases where no default is given.
* Non-values: it is the job of the programmer to filter these out before calling max.  This is very easy to do:
    ```typescript
    const seq = genSequence([undefined, null, NaN, 1, 3, 4]);
    const m = seq
        .filter(v => v != null && v === v)
        .max()
    ```
* one nice addition might be to allow the user to use their own comparison function: `max<T>(gt?: (a: T, b: T) => boolean)`.  This is useful when you wan to find the "oldest" person or the "most expensive" item.